Test failure for compilation utility in macOS 11 and 12

### Failure description

In the sympy test, utilities/_compilation/tests/test_compilation.py::test_compile_link_import_strings is FAILED in macOS.

```
FAILED utilities/_compilation/tests/test_compilation.py::test_compile_link_import_strings - FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/d_...
```

It can be reproduced by
```
py.test utilities/_compilation/tests/test_compilation.py
```
in macOS 11 and 12.

### Environments

The version of SymPy fails the test is `1.11.1` and `1.12.dev`(master branch)

The version of python is (as in macOS 12):
3.8.14(Homebrew) 
```
Python 3.8.14 (default, Sep  6 2022, 23:17:06)
[Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
```

3.9.14(Homebrew)
```
Python 3.9.14 (main, Sep  6 2022, 23:16:16)
[Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
```

3.9.6(XCode Command Line Tools)
```
Python 3.9.6 (default, Aug  5 2022, 15:21:02)
[Clang 14.0.0 (clang-1400.0.29.102)] on darwin
```

The OS version of macOS is:
```
Darwin Kernel Version 21.6.0: Mon Aug 22 20:19:52 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T6000 arm64
```

--- a/.mailmap
+++ b/.mailmap
@@ -746,6 +746,7 @@ Jun Lin <junlin0604@gmail.com> wate123 <junlin0604@gmail.com>
 Jurjen N.E. Bos <jnebos@gmail.com> <jnebos@gmail.com>
 Jurjen N.E. Bos <jnebos@gmail.com> Jurjen N.E. Bos <devnull@localhost>
 Justin Blythe <jblythe29@gmail.com>
+Jyn Spring 琴春 <me@vx.st>
 Ka Yi <chua.kayi@yahoo.com.sg>
 Kaifeng Zhu <cafeeee@gmail.com>
 Kalevi Suominen <jksuom@gmail.com> <kalevi.suominen@helsinki.fi>

--- a/sympy/utilities/_compilation/compilation.py
+++ b/sympy/utilities/_compilation/compilation.py
@@ -215,19 +215,39 @@ def link_py_so(obj_files, so_file=None, cwd=None, libraries=None,
     include_dirs = kwargs.pop('include_dirs', [])
     library_dirs = kwargs.pop('library_dirs', [])
 
-    # from distutils/command/build_ext.py:
+    # Add Python include and library directories
+    # PY_LDFLAGS does not available on all python implementations
+    # e.g. when with pypy, so it's LDFLAGS we need to use
     if sys.platform == "win32":
         warnings.warn("Windows not yet supported.")
     elif sys.platform == 'darwin':
-        # Don't use the default code below
-        pass
+        cfgDict = get_config_vars()
+        kwargs['linkline'] = kwargs.get('linkline', []) + [cfgDict['LDFLAGS']]
+        library_dirs += [cfgDict['LIBDIR']]
+
+        # In macOS, linker needs to compile frameworks
+        # e.g. "-framework CoreFoundation"
+        is_framework = False
+        for opt in cfgDict['LIBS'].split():
+            if is_framework:
+                kwargs['linkline'] = kwargs.get('linkline', []) + ['-framework', opt]
+                is_framework = False
+            elif opt.startswith('-l'):
+                libraries.append(opt[2:])
+            elif opt.startswith('-framework'):
+                is_framework = True
+        # The python library is not included in LIBS
+        libfile = cfgDict['LIBRARY']
+        libname = ".".join(libfile.split('.')[:-1])[3:]
+        libraries.append(libname)
+
     elif sys.platform[:3] == 'aix':
         # Don't use the default code below
         pass
     else:
         if get_config_var('Py_ENABLE_SHARED'):
             cfgDict = get_config_vars()
-            kwargs['linkline'] = kwargs.get('linkline', []) + [cfgDict['PY_LDFLAGS']] # PY_LDFLAGS or just LDFLAGS?
+            kwargs['linkline'] = kwargs.get('linkline', []) + [cfgDict['LDFLAGS']]
             library_dirs += [cfgDict['LIBDIR']]
             for opt in cfgDict['BLDLIBRARY'].split():
                 if opt.startswith('-l'):
@@ -291,7 +311,11 @@ def simple_cythonize(src, destdir=None, cwd=None, **cy_kwargs):
         cy_result = cy_compile([src], cy_options)
         if cy_result.num_errors > 0:
             raise ValueError("Cython compilation failed.")
-        if os.path.abspath(os.path.dirname(src)) != os.path.abspath(destdir):
+
+        # Move generated C file to destination
+        # In macOS, the generated C file is in the same directory as the source
+        # but the /var is a symlink to /private/var, so we need to use realpath
+        if os.path.realpath(os.path.dirname(src)) != os.path.realpath(destdir):
             if os.path.exists(dstfile):
                 os.unlink(dstfile)
             shutil.move(os.path.join(os.path.dirname(src), c_name), destdir)

